Add Variant marshalling

kant2002 opened the pull request:
#1100, #306

kant2002 committed:
Add Variant marshalling

kant2002 commented:
I did not test that yet. Just trying to gather feedback on the direction of the change.

kant2002 committed:
Implement cleanup

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs:
```diff
@@ -126,9 +146,15 @@ public static object GetObjectForIUnknown(IntPtr pUnk)
+            }
+
+            Variant* data = (Variant*)pSrcNativeVariant;
+            return data->ToObject();
```
This is going to call `GetObjectForNativeVariant` in some cases that will lead to infinite recursion.

jkotas replied:
I am not sure why `ToObject` handles some of the cases inline and does not just call `GetObjectForNativeVariant` all the time.

kant2002 replied:
Upon more investigation I would think I incorrectly use `Variant.CopyFromIndirect`. Only place where I see it's usage is in https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/Common/src/System/Runtime/InteropServices/ComEventsSink.cs#L231 
I thought I do smart, move, apparently I have to reimplement `GetObjectForNativeVariant` in similar manner.

Also I would like to ask @jkotas what should I do with test failing due to reflection. I assume reflection needed for type checks `is typeof(byte)`. I do not see a way how to solve that, maybe move test to separate class where reflection would not be blocked? Please advice.

jkotas replied:
`is typeof(byte)` does not use reflection. 

The tests failures are caused by invalid IL produced by the new marshaller that you have added.

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs:
```diff
@@ -102,15 +102,35 @@ public static IntPtr GetIUnknownForObject(object o)
+                return;
+            }
+
+            if (obj is byte)
```
This should handle the same set of cases as the other direction.

kant2002 replied:
I was not sure. You may have idea in what order should I check each type based on usage. That was just a sketch.

kant2002 commented:
With half backed implementation I hit this

![image](https://user-images.githubusercontent.com/4257079/119001488-d2ad8500-b9ad-11eb-99f9-1fd44142299f.png)

Seems to be IDispatch is inevitable. Will need advice after this would be done.

kant2002 committed:
Not tested marshalling
- Marshal_XXX_Boolean
- Marshal_XXX_Currency
- Marshal_XXX_Object
- Marshal_XXX_Object_IUnknown

VariantBool somehow does not recognized. Probably something is messed up. I will do that in subsequent commit.

kant2002 commented:
Not tested
- Marshal_XXX_Boolean
- Marshal_XXX_Currency
- Marshal_XXX_Object
- Marshal_XXX_Object_IUnknown

Current tests in the `Interop\Pinvoke\Variant` depends on the built-in marshalling. Is it possible split tests that built-in marshalling would be not required. Or maybe I can create similar set of test for ComWrappers code path which will share most of code with current tests by reusing files? Do I need go to runtime for that?

jkotas commented:
> Is it possible split tests that built-in marshalling would be not required.

Yes. Note that there is ongoing work to split the tests that require built-in COM marshalling in dotnet/runtime, so contributing to that would be definitely welcomed:

https://github.com/dotnet/runtime/issues/50500
https://github.com/dotnet/runtime/issues/51265

jkotas commented:
> Seems to be IDispatch is inevitable. Will need advice after this would be done.

Does this really only need to QueryInterface for `IID_IDispatch`?

kant2002 commented:
> Does this really only need to QueryInterface for `IID_IDispatch`?

Yes. Pedantic me want to implement `Marshal.GetObjectForComInterface<object, T>` which is overkill or this specific case. If I change Variant implementation in runtime to just QI to IDispatch that should just works.

AaronRobinsonMSFT commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs:
```diff
@@ -102,15 +102,105 @@ public static IntPtr GetIUnknownForObject(object o)
+            }
+            else if (obj is System.Reflection.Missing)
+            {
+                data->AsError = -2147352572 /*DISP_E_PARAMNOTFOUND*/;
```
Hex is preferred for `HRESULT`s.

AaronRobinsonMSFT commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:
```diff
@@ -1037,4 +1041,72 @@ protected override void TransformNativeToManaged(ILCodeStream codeStream)
+    {
+        protected override void AllocAndTransformManagedToNative(ILCodeStream codeStream)
+        {
+            LoadManagedValue(codeStream);
```
`VARIANT_BOOL` marshalling is a bit more complicated. See https://github.com/dotnet/runtime/blob/aaaaeed8c8d2d01dccd1f7fc536ea27eee75f472/src/coreclr/vm/ilmarshalers.cpp#L195-L221 and https://github.com/dotnet/runtime/blob/aaaaeed8c8d2d01dccd1f7fc536ea27eee75f472/src/coreclr/vm/ilmarshalers.h#L1960-L1987

kant2002 committed:
Add bool and OLE currency marshalling

kant2002 committed:
Add missing interop

kant2002 committed:
Use constant for DISP_E_PARAMNOTFOUND

kant2002 commented:
Locally seems to be working. Build issues seems to be unrelated. But current PR I rely on the changes to runtime which is not yet landed, and as such not integrated.

ComboBox now demands IDispatch implementation from my ComWrappers instance. Sigh.
Will try to use https://github.com/dotnet/samples/blob/main/core/interop/comwrappers/IDispatch/AnyObjectProxy.cs as example how to do that. But overall work on the Variant marshalling seems to be done in technical sense. Ready to accept feedback.

kant2002 committed:
Add couple rare cases

kant2002 commented:
Actually three lines may not working

https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/tests/Interop/PInvoke/Variant/VariantTest.cs#L46-L48

It's not clear for me 
- First line means that TimeSpan not marshalled, or specifically TimeSpan.Zero cannot be marshalled.
- Third line VariantWrapper not supported at all? Actually I found https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.variantwrapper?view=net-5.0#remarks but have not incentive to implement that now because I do not fully understand what's going on here.
- Second line. Does any not implemented cases should throw `NotSupportedException`?

kant2002 committed:
Fix compilation error

jkotas commented:
> First line means that TimeSpan not marshalled, or specifically TimeSpan.Zero cannot be marshalled.

Yeah, that looks odd. It would be useful to run it on CoreCLR to see where it is comming from.

> Does any not implemented cases should throw NotSupportedException?

It should throw whatever CoreCLR throws. In general, there is no much consistency around `NotSupportedException` and similar exceptions - see https://github.com/dotnet/runtime/issues/22069#issuecomment-831563723 for some stats.

The full Variant marshalling is very complex. It is fine to only implement a reasonable subset.

kant2002 commented:
> The full Variant marshalling is very complex. It is fine to only implement a reasonable subset.

Can we leave this 3 cases? I would check TimeSpan for sure, but would like to move forward if possible

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs:
```diff
@@ -68,7 +71,17 @@ public static int FinalReleaseComObject(object o)
+                throw new ArgumentNullException(nameof(T));
+            }
+
+            return ComWrappers.ComInterfaceForObject(o, new Guid(T.GetCustomAttribute<GuidAttribute>().Value));
```
Why not just call `Type.GUID` here? `GetCustomAttribute` is not going to be any faster than `Type.GUID`.

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System.Private.CoreLib.csproj:
```diff
@@ -257,6 +257,12 @@
   <ItemGroup Condition="'$(TargetsWindows)'=='true'">
     <Compile Include="Microsoft\Win32\SafeHandles\SafeThreadPoolIOHandle.cs" />
     <Compile Include="System\Runtime\InteropServices\PInvokeMarshal.Windows.cs" />
+    <Compile Include="$(CommonPath)\System\Runtime\InteropServices\IDispatch.cs">
```
This should not be necessary once your dotnet/runtime change gets in. Is that rigth? It may be better to cherry pick that change into this PR and delete this line.

jkotas commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs:
```diff
@@ -1405,13 +1407,40 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
     class BooleanMarshaller : Marshaller
     {
+        private int _trueValue;
+        private int _falseValue;
```
Nit: It is not necessary to have falseValue configurable. It is always 0.

jkotas replied:
And `AllocAndTransformNativeToManaged` assumes that the falseValue is zero.

kant2002 committed:
Use Type.GUID instead of querying for custom attributes

kant2002 committed:
Remove constant false value

kant2002 committed:
Remove no longer needed interface

kant2002 commented:
@jkotas build issues seems to be unrelated. Do I miss anything else here?

jkotas commented:
It is failing with:
```
2021-05-24T09:16:57.1984007Z   Assertion failed 'genActualType(lclTyp) == genActualType(op1->gtType) || (genActualType(lclTyp) == TYP_I_IMPL && op1->IsLocalAddrExpr() != nullptr) || (genActualType(lclTyp) == TYP_I_IMPL && (op1->gtType == TYP_BYREF || op1->gtType == TYP_REF)) || (genActualType(op1->gtType) == TYP_I_IMPL && lclTyp == TYP_BYREF) || (varTypeIsFloating(lclTyp) && varTypeIsFloating(op1->TypeGet())) || ((genActualType(lclTyp) == TYP_BYREF) && genActualType(op1->TypeGet()) == TYP_REF) : Possibly bad IL with CEE_stloc.s at offset 0013h (op1=ref op2=NULL stkDepth=0)' in 'Internal.CompilerGenerated.<Module>:ReverseDelegateStub__Get(int,long,long,int,long,long,long):int' during 'Importation' (IL size 84)
```
It looks like bad IL generated by the new marshaler.

kant2002 commented:
@jkotas this was fixed in https://github.com/dotnet/runtimelab/pull/1102/commits/ea76dd01f7060e4b5f486ffd7468746170cd9211 as part of #1102 . Exception was part of `System.Management.WmiNetUtilsHelper.GetNames` 

```
internal delegate int GetNames(
  int vFunc,
  IntPtr pWbemClassObject,
  [In][MarshalAs(UnmanagedType.LPWStr)] string wszQualifierName,
  [In] int lFlags,
  [In] ref object pQualifierVal, 
->>  [MarshalAs(UnmanagedType.SafeArray, SafeArraySubType = VarEnum.VT_BSTR)] out string[] pNames);
```

kant2002 commented:
Other location is 
```
// System.Management.WmiNetUtilsHelper.GetDemultiplexedStub

internal delegate int GetDemultiplexedStub(
  [In][MarshalAs(UnmanagedType.IUnknown)] object pIUnknown,
  [In] bool isLocal, 
-->  [MarshalAs(UnmanagedType.IUnknown)] out object ppIUnknown);
```

For me this seems as unrelated change. (This one actually exposes another issue with IUnknown marshalling). I need more time to understand what's going on.

kant2002 commented:
So based on my understanding, problem is with delegates who has `[In][Out]ref object` as the parameter.
Current implementation of marshaller does not handle that. And seems I would like somehow to opt-out of that temporary, but do not know how. If possible and somebody know any hints, I would appreciate.

This indicates missing coverage in tests. If I have courage, I will find out what case trigger issue, then I may create test. Should I do that, if so, where should I do that?

jkotas commented:
Throw exception for the unhandled case before it crashes the JIT.

kant2002 committed:
Do not support reverse direction for marshalling of VARIANT

kant2002 commented:
Phew, I'm finally understand where to put exception. My understanding that this was related to https://github.com/dotnet/runtimelab/issues/176 because I was able to replicate with 
```
Get xxx = Marshal.GetDelegateForFunctionPointer<Get>(IntPtr.Zero);
internal delegate int Get(ref object pVal);
```

kant2002 commented:
@jkotas anything here?

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs:
```diff
@@ -12,6 +12,9 @@ namespace System.Runtime.InteropServices
 {
     public static partial class Marshal
     {
+        private static readonly Guid IID_IDispatch = new Guid("00020400-0000-0000-C000-000000000046");
```
This will be part of the Marshal static constructor and thus this will run and the code (will all dependencies) be included in the binary even when the app is using Marshal, but not using  `GetIDispatchForObject` at all. I would instead inline the guid into GetIDispatchForObject

`return ComWrappers.ComInterfaceForObject(o, new Guid(0x00020400, ...) /* IID_IDispatch */);

jkotas commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs:
```diff
@@ -1037,4 +1043,86 @@ protected override void TransformNativeToManaged(ILCodeStream codeStream)
+
+        protected override void EmitCleanupManaged(ILCodeStream codeStream)
+        {
+            // Only do cleanup if it is IN
```
This comment should be other way around. This does cleanup only if it is not IN

kant2002 replied:
This is actually fine comment. We exit when `not IN`, so we `do cleanup` when IN. 

Interesting observation, that  negation is complicated for humans. I'm kind of interested in psychological properties of linguistic, so cannot resist take a note.

jkotas replied:
Ah ok.

kant2002 committed:
Address PR feedback

jkotas commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.cs:
```diff
@@ -1411,13 +1413,41 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
+                codeStream.EmitLdc(_trueValue);
+                codeStream.Emit(ILOpcode.br, pDoneLabel);
+#if DEBUG
+                codeStream.Emit(ILOpcode.pop);   // keep the simple stack level calculator happy
```
Is this really needed? I see number of other places in this file that use similar pattern and they do not need this workaround.

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs:
```diff
@@ -127,9 +260,16 @@ public static object GetObjectForIUnknown(IntPtr pUnk)
+        public static unsafe object? GetObjectForNativeVariant(IntPtr pSrcNativeVariant)
         {
-            throw new PlatformNotSupportedException(SR.PlatformNotSupported_ComInterop);
+            if (pSrcNativeVariant == IntPtr.Zero)
```
CoreCLR implementation throws for this case.

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs:
```diff
@@ -127,9 +260,16 @@ public static object GetObjectForIUnknown(IntPtr pUnk)
+
+            Variant* data = (Variant*)pSrcNativeVariant;
+
+            return data->ToObject();
```
I think that this has still the problem of hitting infinite recursion. For example, `ToObject` will call `GetObjectForNativeVariant` for `VT_RECORD` variant.

jkotas replied:
I am ok with not supporting the less common variant types like ``VT_RECORD`, but it needs to throw exception instead of infinite recursion.

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs:
```diff
@@ -103,15 +120,131 @@ public static IntPtr GetIUnknownForObject(object o)
         [SupportedOSPlatform("windows")]
-        public static void GetNativeVariantForObject(object? obj, IntPtr pDstNativeVariant)
+        public static unsafe void GetNativeVariantForObject(object? obj, IntPtr pDstNativeVariant)
         {
```
The CoreCLR impl checks for `pDstNativeVariant == NULL`

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs:
```diff
@@ -103,15 +120,131 @@ public static IntPtr GetIUnknownForObject(object o)
+                }
+                else if (type.IsArray)
+                {
+                    // SAFEARRAY impementation goes here.
```
Typo: impementation

kant2002 committed:
Apply suggestions from feedback

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs:
```diff
@@ -103,15 +120,136 @@ public static IntPtr GetIUnknownForObject(object o)
+                return;
+            }
+
+            if (obj is bool flag)
```
This would look better with pattern matching:
```
switch (obj)
{
case bool value:
    data->AsBool = value;
    break;
case byte value;
    data->AsUi1 = value;
    break;
...
default:
...
}
```
Also, I would put int and string first since they are the most frequently used variants (the CoreCLR unmanaged impl has them first).

kant2002 committed:
Stylistical change of multiple ifs to switch.

jkotas reviewed (approved):
LGTM. Thank you!

kant2002 committed:
Fix typo
